feat: add 'pre-commit' and 'git-crypt' to workflow

sowecthal opened the pull request:
### Summary: 

**1. Install tools**

- https://pre-commit.com/
- https://github.com/AGWA/git-crypt

**2. Add `pre-commit` hooks to the repository (including the following rules)** 

- https://github.com/adrienverge/yamllint.git
- https://github.com/lyz-code/yamlfix/
- https://github.com/shellcheck-py/shellcheck-py
- https://github.com/petalmd/git-crypt-pre-commit (ensure git-crypt is installed)
- https://github.com/pre-commit/pre-commit-hooks (check-added-large-files, check-merge-conflict, forbid-new-submodules, no-commit-to-branch, check-symlinks, mixed-line-ending, end-of-file-fixer, trailing-whitespace)

**3. Reorganize the handling of secrets.**

Adjust the test task to take secret parameters from an encrypted file (`secrets.txt` encrypted with `git-crypt`) instead of using GitHub secrets.

**4. Open a new PR**

kseniyashaydurova commented on .git-crypt/.gitattributes:
Let's add whole content of `.git-crypt` folder to `.gitignore`, as well as below ones:
- .idea
- .vscode
- git-crypt-key

sowecthal replied:
Done

kseniyashaydurova commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,30 @@
+    rev: v1.35.1
+    hooks:
+      - id: yamllint
+        entry: yamllint -c .yamllint.yaml .
```
Is it really needed?

sowecthal replied:
Fixed

kseniyashaydurova commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,30 @@
+    rev: 1.17.0
+    hooks:
+      - id: yamlfix
+  - repo: https://github.com/shellcheck-py/shellcheck-py
```
Let's find a param that we can pass to shellcheck to check also attached `sources` to bash files - https://github.com/koalaman/shellcheck

sowecthal replied:
Done

kseniyashaydurova commented on .pre-commit-config.yaml:
LEt's order all hooks and repos in alphabetical order

sowecthal replied:
Done

kseniyashaydurova commented on .git-crypt/.gitattributes:
Let's add whole content of `.git-crypt` folder to `.gitignore`, as well as below ones:
- .idea
- .vscode
- git-crypt-key

sowecthal replied:
Done

kseniyashaydurova commented on .github/workflows/ci.yaml:
```diff
@@ -66,61 +59,65 @@ jobs:
+        run: echo "execution_time=$(gh run list --json createdAt,updatedAt -w ci.yaml
+          -L 1 | jq -r '.[0] | (( (strtotime(.updatedAt) - strtotime(.createdAt))
+          ))')s" >> $GITHUB_ENV
+      - name: Unlock secrets with git-crypt
```
Pls specify your yamllint and yamlfix so it allows to separate steps with empty lines

sowecthal replied:
Fixed

kseniyashaydurova commented on .github/workflows/ci.yaml:
```diff
@@ -66,61 +59,65 @@ jobs:
+        id: read-secrets
+        if: always()
+        run: |
+          cat secrets.txt >> "$GITHUB_OUTPUT"
```
I don't see that you use anywhere vars from your secrets in this CI/CD

sowecthal replied:
Fixed

kseniyashaydurova commented on .github/workflows/ci.yaml:
```diff
@@ -43,20 +43,13 @@ on:
         default: Github actions job result
         type: string
     secrets:
-      mail_username:
```
Why they are removed? You still use them

sowecthal replied:
Fixed

kseniyashaydurova commented on .github/workflows/ci.yaml:
```diff
@@ -66,61 +59,65 @@ jobs:
           server_port: ${{ inputs.mail_server_port }}
-          username: ${{ secrets.mail_username }}
-          password: ${{ secrets.mail_password }}
+          username: ${{ secrets.MAIL_USERNAME }}
```
Why you use now secrets in higher case?

sowecthal replied:
Fixed

kseniyashaydurova commented on .yamllint.yaml:
```diff
@@ -0,0 +1,8 @@
+    max: 100
+    level: warning
+#   # don't bother me with this rule
+#   indentation: disable
```
Why commented?

sowecthal replied:
Fixed

kseniyashaydurova commented on secrets.txt:
```diff
@@ -0,0 +1,2 @@
+MAIL_USERNAME=sowecthal@gmail.com
```
Why secrets are not encrypted? Now anybody would be able to use your password. Pls reorganize your git histroy to not have such unencrypted secrets anywhere in git commits

sowecthal replied:
Fixed

kseniyashaydurova reviewed (changes requested):
@sowecthal pls:
1. Squash big commits count prior opening all PRs to single one
2. Encrypt secrets
3. Do like we discussed with Ilya. located only `MAIL_SERVER` and `MAIL_SERVER_PORT` and other not sensitive secrets in our encrypted secrets file, leave your email and password in GH secrets (cause the action we use for sending emails will show it in github action logs in open text)
4. Clean up all your builds history too, where ithere are printed your google creds

sowecthal committed:
Chore: Squashed all commits into one

Experimental commit. Added secrets.txt

Experimental commit. Added debug

Experimental commit. Removed debug

kseniyashaydurova commented on .github/workflows/ci.yaml:
```diff
@@ -77,16 +69,16 @@ jobs:
         run: npm run build
 
-      - name: Log in to github container registry  
+      - name: Log in to github container registry
```
LEt's find out existing actionfor docker login

sowecthal replied:
Done

kseniyashaydurova commented on .github/workflows/ci.yaml:
```diff
@@ -77,16 +69,16 @@ jobs:
+          echo "${{ github.token }}" \
           | docker login ${{ inputs.docker_registry }} -u ${{ github.actor }} --password-stdin
 
       - name: Build docker image
```
Let's find out existing actions for docker `build` and `push`

sowecthal replied:
Done

kseniyashaydurova commented on .github/workflows/ci.yaml:
```diff
@@ -103,24 +95,24 @@ jobs:
           password: ${{ secrets.mail_password }}
           subject: ${{ inputs.mail_subject  }}
-          to: ${{ inputs.mail_receiver != '' && inputs.mail_receiver || env.DEFAULT_MAIL_RECEIVER }} 
+          to: ${{ secrets.mail_to != '' && secrets.mail_to || env.DEFAULT_MAIL_RECEIVER }}
```
Can you replace to this
```suggestion
          to: ${{ secrets.mail_to || env.DEFAULT_MAIL_RECEIVER }}
```

sowecthal replied:
Done

kseniyashaydurova commented on .github/workflows/get-pass-secrets.yaml:
```diff
@@ -0,0 +1,49 @@
+name: Get and pass secrets
```
Bad name for file and workflow itself, basically you don't only create and pass secrets here, you perform whole CI/CD

sowecthal replied:
Fixed

kseniyashaydurova commented on .github/workflows/get-pass-secrets.yaml:
```diff
@@ -0,0 +1,49 @@
+          cat secrets.txt >> "$GITHUB_OUTPUT"
+        working-directory: .
+
+  pass-secrets:
```
Bad name for job, should be `build`

sowecthal replied:
Fixed

kseniyashaydurova commented on .github/workflows/get-pass-secrets.yaml:
```diff
@@ -0,0 +1,49 @@
+        uses: actions/checkout@v4
+
+      - name: Unlock secrets with git-crypt
+        if: always()
```
Is it really needed?

sowecthal replied:
Fixed

kseniyashaydurova commented on .github/workflows/get-pass-secrets.yaml:
```diff
@@ -0,0 +1,49 @@
+
+      - name: Read secrets from secrets.txt
+        id: read-secrets
+        if: always()
```
Is it really needed?

sowecthal replied:
Fixed

kseniyashaydurova commented on .github/workflows/get-pass-secrets.yaml:
```diff
@@ -0,0 +1,49 @@
+        if: always()
+        run: |
+          cat secrets.txt >> "$GITHUB_OUTPUT"
+        working-directory: .
```
Is it really needed? Let's specify it as default params?

sowecthal replied:
Fixed

kseniyashaydurova commented on .github/workflows/get-pass-secrets.yaml:
```diff
@@ -0,0 +1,49 @@
+    needs: get-secrets
+    uses: ./.github/workflows/ci.yaml
+    with:
+      node_version: 18
```
LEt's update script to dynamically get node version from package.json (check how you can do it with existing GH actions)

sowecthal replied:
Done

kseniyashaydurova commented on .github/workflows/ci.yaml:
LEt's name this file `build`

sowecthal replied:
Done

kseniyashaydurova commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,34 @@
+    rev: v0.10.0.1
+    hooks:
+      - id: shellcheck
+        entry: shellcheck -x
```
Pls try to do the same with `args`

sowecthal replied:
Done

kseniyashaydurova commented on .yamllint.yaml:
```diff
@@ -0,0 +1,5 @@
+extends: default
```
Pls use this config:

```
extends: default

yaml-files:
  - '*.yaml'
  - '*.yml'
  - '.yamllint'

# see rules
# https://yamllint.readthedocs.io/en/stable/rules.html
rules:
  line-length: disable
  document-start: disable
  comments: disable
```

sowecthal replied:
Done

kseniyashaydurova reviewed (changes requested):
@sowecthal pls add few more improvements :)

sowecthal committed:
Chore: Made changes according to review

Chore: Used an action for docker build and push

Chore: Modified the step for docker build and push

Chore: Modified the trigger workflow to match docker build and push action

Chore: Changed the recipient line in the send mail step

Chore: Renamed workflows and jobs

Chore: Got rid of unnecessary lines

Chore: Updated yamllint config

Chore: Added args to shellcheck in the pre-commit config

Chore: Updated script to dynamically get node version

Chore: Fixed a minor mistake
